tests/agreement.rs

```diff
@@ -0,0 +1,169 @@
+}
+
+impl TestNetwork {
+    fn new(num_nodes: usize) -> TestNetwork {
+        // Make a node with an Agreement instance associated with the proposer node 0.
+        let make_node = |id: NodeId| (id, TestNode::new(id, Agreement::new(NodeId(0), num_nodes)));
+        TestNetwork {
+            nodes: (0..num_nodes).map(NodeId).map(make_node).collect(),
+            scheduled_node_id: NodeId(0),
+        }
+    }
+
+    fn dispatch_messages(&mut self, sender_id: NodeId, messages: InstanceQueue) {
+        for message in messages {
+            for (id, node) in &mut self.nodes {
+                if *id != sender_id {
+                    debug!(
+                        "Dispatching from {:?} to {:?}: {:?}",
+                        sender_id, id, message
+                    );
+                    node.queue.push_back((sender_id, message.clone()));
+                }
+            }
+        }
+    }
+
+    // Gets a node for receiving a message and picks the next node with a
+    // non-empty message queue in a cyclic order.
+    fn pick_node(&mut self) -> NodeId {
+        let id = self.scheduled_node_id;
+        // Try a node with a higher ID for fairness.
+        if let Some(next_id) = self.nodes
+            .iter()
+            .find(|(&next_id, node)| id < next_id && !node.queue.is_empty())
+            .map(|(id, _)| *id)
+        {
+            self.scheduled_node_id = next_id;
+        } else {
+            // Fall back to nodes up to the currently scheduled ID.
+            self.scheduled_node_id = self.nodes
+                .iter()
+                .find(|(&next_id, node)| id >= next_id && !node.queue.is_empty())
+                .map(|(id, _)| *id)
+                .expect("no more messages in any node's queue")
+        }
+        debug!("Picked node {:?}", self.scheduled_node_id);
+        id
+    }
+
+    fn step(&mut self) -> (NodeId, Option<bool>) {
+        let sender_id = self.pick_node();
+        let (output, messages) = self.nodes.get_mut(&sender_id).unwrap().handle_message();
+        self.dispatch_messages(sender_id, messages);
+        (sender_id, output)
+    }
+
+    fn set_input(&mut self, sender_id: NodeId, input: bool) {
+        let message = self.nodes
+            .get_mut(&sender_id)
+            .unwrap()
+            .agreement
+            .set_input(input)
+            .expect("set input");
+        self.dispatch_messages(sender_id, VecDeque::from(vec![message]));
+    }
+}
+
+fn test_agreement(mut network: TestNetwork) -> BTreeMap<NodeId, TestNode> {
+    let _ = env_logger::try_init();
+
+    // Pick the first node with a non-empty queue.
+    network.pick_node();
+
+    while network.nodes.values().any(|node| node.outputs.is_empty()) {
+        let (NodeId(id), output) = network.step();
+        if let Some(value) = output {
+            debug!("Node {} output {}", id, value);
+        }
+    }
+    network.nodes
+}
+
+/// Test 3 correct and 1 faulty node. The faulty node simply negates all other
```

Comment:
Node 3 isn't really faulty, it just starts with a different value but follows the protocol correctly.